File: assign4/javascript/src/guess_game.js
'use strict';

const _ = require('lodash');
const seedrandom = require('seedrandom');

const GuessResult = {
  LOW: 'LOW',
  CORRECT: 'CORRECT',
  HIGH: 'HIGH',
};

function guess(target, guess) {
  return [GuessResult.LOW, GuessResult.CORRECT, GuessResult.HIGH][Math.sign(guess - target) + 1];
}

function playGame(target, readGuess, display) {
  return _(_.range(1, 1000))
    .map(attempt => [attempt, readGuess()])
    .map(([attempt, aGuess]) => [attempt, guess(target, aGuess)])
    .map(([attempt, result]) => [attempt, display(result) || result])
    .find(([attempt, result]) => result === GuessResult.CORRECT)[0];
}

function getRandomNumber(seed) {
  return Math.floor(seedrandom(seed)() * 100) + 1;
}

module.exports = { guess, playGame, getRandomNumber, GuessResult };

File: assign4/f#/src/guess_game.fs
namespace src

open System

module GuessGame =
  type GuessResult =
    | LOW
    | HIGH 
    | CORRECT

  let guess (target: int) (guess: int) = 
    [|GuessResult.LOW; GuessResult.CORRECT; GuessResult.HIGH|].[Math.Sign(guess - target) + 1]

  let playGame (target: int, readGuess: unit -> int, display: GuessResult -> unit) : int =
    Seq.initInfinite ((+) 1)
    |> Seq.map (fun attempt -> attempt, readGuess())
    |> Seq.map (fun (attempt, userGuess) -> attempt, (guess target userGuess))
    |> Seq.map (fun (attempt, result) -> display result; (attempt, result))
    |> Seq.find (fun (_, result) -> result = GuessResult.CORRECT)
    |> fst

  let getRandomNumber (seed: int) =
    Random(seed).Next(1, 101)
    